javascript/ql/src/semmle/javascript/frameworks/NodeJSLib.qll

```diff
@@ -944,9 +954,96 @@ module NodeJSLib {
   /**
    * A dispatch of an event on a NodeJS EventEmitter instance.
    */
-  private class NodeJSEventDispatch extends EventDispatch::DefaultEventDispatch, DataFlow::MethodCallNode {
+  private class NodeJSEventDispatch extends EventDispatch::DefaultEventDispatch,
+    DataFlow::MethodCallNode {
     override NodeJSEventEmitter emitter;
 
     NodeJSEventDispatch() { this = emitter.ref().getAMethodCall("emit") }
+    
+    override EventRegistration::Range getAReceiver() { emitter.getBaseEmitter() = result.getEmitter().(NodeJSEventEmitter).getBaseEmitter() }
```

Comment:
If I do your change above related to `ref()` and `getAnInstanceReference`, then it might be easy to achieve.

Reply:
Your suggestion of changing `ref()` worked out nicely. 

> Instead of treating all instances of the same class and aliased, it would be nice if we could achieve the following:
> 
> Event dispatched on this can be received by all instances and subclasses
> Event received on this can come from all instances and subclasses
> Events dispatched on one instance or subclass cannot be dispatched to another instance or subclass

With the change to the `.ref()` method we have this now.